pkg/upgrade/kubeconfig.go

```diff
@@ -0,0 +1,231 @@
+package upgrade
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"strings"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/certificates" // RIP
+	clusterapiv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+type secretsClient interface {
+	Get(string, metav1.GetOptions) (*v1.Secret, error)
+}
+type decoder interface {
+	Decode([]byte, []byte) (int, error)
+	DecodedLen(n int) int
+}
+type kubeconfigClient interface {
+	Write(clientcmdapi.Config) ([]byte, error)
+	RESTConfigFromKubeConfig([]byte) (*restclient.Config, error)
+	NewKubeconfig(string, string, *x509.Certificate, *rsa.PrivateKey) (*clientcmdapi.Config, error)
+	DecodeCertPEM([]byte) (*x509.Certificate, error)
+	DecodePrivateKeyPEM([]byte) (*rsa.PrivateKey, error)
+}
+
+type kubeconfig struct{}
+
+func (k *kubeconfig) NewKubeconfig(a, b string, c *x509.Certificate, d *rsa.PrivateKey) (*clientcmdapi.Config, error) {
+	return certificates.NewKubeconfig(a, b, c, d)
+}
+func (k *kubeconfig) RESTConfigFromKubeConfig(a []byte) (*restclient.Config, error) {
+	return clientcmd.RESTConfigFromKubeConfig(a)
+}
+func (k *kubeconfig) Write(a clientcmdapi.Config) ([]byte, error) {
+	return clientcmd.Write(a)
+}
+func (k *kubeconfig) DecodeCertPEM(a []byte) (*x509.Certificate, error) {
+	return certificates.DecodeCertPEM(a)
+}
+func (k *kubeconfig) DecodePrivateKeyPEM(a []byte) (*rsa.PrivateKey, error) {
+	return certificates.DecodePrivateKeyPEM(a)
+}
+
+type unstructuredFunctions interface {
+	ToUnstructured(interface{}) (map[string]interface{}, error)
+	NestedString(map[string]interface{}, ...string) (string, bool, error)
+}
+type unstructuredClient struct{}
+
+func (u *unstructuredClient) ToUnstructured(i interface{}) (map[string]interface{}, error) {
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(i)
+}
+func (u *unstructuredClient) NestedString(obj map[string]interface{}, fields ...string) (string, bool, error) {
+	return unstructured.NestedString(obj, fields...)
+}
+
+// keyPair contains a cert and key.
+type keyPair struct {
+	Cert []byte `json:"cert"`
+	Key  []byte `json:"key"`
+}
+
+type keyPairGetter struct {
+	secretsClient      secretsClient
+	decoder            decoder
+	kubeconfigClient   kubeconfigClient
+	unstructuredClient unstructuredFunctions
+}
+
+// GetRestConfig figures out how to build a rest.Config from one of three possible ways based on the KeyPairConfig passed in.
+func GetRestConfig(secretsClient secretsClient, decoder decoder, kubeconfigClient kubeconfigClient, cluster *clusterapiv1alpha1.Cluster, targetClusterAPIEndpoint string, config KeyPairConfig) (*rest.Config, error) {
```

Comment:
Before I go any further on the review, I have one main question: now that we have 2 different approaches to getting a rest config, do you think it's still worth it to consolidate all this logic into a single function? I'm finding the signature (6 parameters) to be unwieldy. Do you think there's a way to make this cleaner?